openstax/middleware.py

```diff
@@ -39,3 +48,95 @@ def process_response(self, request, response):
             response = self.handler.get_response(request)
 
         return response
+
+
+class CommonMiddlewareOpenGraphRedirect(CommonMiddleware):
+    OG_USER_AGENTS = [
+        'twitterbot',
+        'facebookbot',
+        'pinterestbot',
+        'slackbot-linkexpanding',
+    ]
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request, *args, **kwargs):
+        if 'HTTP_USER_AGENT' in request.META:
+
+            user_agent = user_agent_parser.ParseUserAgent(request.META["HTTP_USER_AGENT"])
+            if user_agent['family'].lower() in self.OG_USER_AGENTS:
+                # url path minus the trailing /
+                url_path = request.get_full_path()[:-1]
+
+                full_url = request.build_absolute_uri()
+
+                # imdex of last / to find slug, except when there isn't a last /
+                if url_path == '':
+                    page_slug = "openstax-homepage"
+                else:
+                    index = url_path.rindex('/')
+                    page_slug = url_path[index+1:]
+
+                if self.redirect_path_found(url_path):
+                    # supporters page has the wrong slug
+                    if page_slug == 'foundation':
+                        page_slug = 'supporters'
+
+                    # look up correct object based on path
+                    if '/details/books/' in url_path:
+                        page = Book.objects.filter(slug = page_slug)
+                    elif '/blog/' in url_path:
+                        page = NewsArticle.objects.filter(slug = page_slug)
+                    else:
+                        page = self.page_by_slug(page_slug)
+
+                    template = self.build_template(page[0], full_url)
+                    return HttpResponse(template)
+                else:
+                    return self.get_response(request)
+        return self.get_response(request)
+
+    def build_template(self, page, page_url):
+        image_url = self.image_url(page.promote_image)
+        template = '<!DOCTYPE html> <html> <head> <meta charset="utf-8">'
+        template += '<title>' + str(page.seo_title) + '</title>'
+        template += '<meta name = "description" content = "{}" >'.format(page.search_description)
+        template += '<link rel = "canonical" href = "{}" />'.format(page_url)
+        template += '<meta name = "og:url" content = "{}" >'.format(page_url)
+        template += '<meta property="og:type" content="article" />'
+        template += '<meta name = "og:title" content = "{}" >'.format(page.seo_title)
```

Comment:
some of these have closing HTML tags (see line above), some don't - not sure if it matters

Reply:
I took the HTML from the existing template page.html. I left it as is since the template works.